adapter.go

```diff
@@ -384,7 +460,40 @@ func (a *Adapter) getRange(key string, rng blockRange) ([][]byte, error) {
 			nToFetch++
 		}
 	}
-	if nToFetch == len(blocks) {
+	if nToFetch == len(blocks) && a.keyStreamer != nil {
+		r, err := a.srcStreamAt(key, rng.start*a.blockSize, (rng.end-rng.start+1)*a.blockSize)
+		if err != nil {
+			for i := rng.start; i <= rng.end; i++ {
+				blockID := a.blockKey(key, i)
+				a.blmu.Unlock(blockID)
+			}
+			return nil, err
+		}
+		defer r.Close()
+		for bid := int64(0); bid <= rng.end-rng.start; bid++ {
+			blockID := a.blockKey(key, bid+rng.start)
+			buf := make([]byte, a.blockSize)
+			n, err := io.ReadFull(r, buf)
+			if err == io.ErrUnexpectedEOF {
+				err = io.EOF
+			}
+			if err == nil || err == io.EOF {
+				blocks[bid] = buf[:n]
+				a.cache.Add(key, uint(rng.start+bid), blocks[bid])
+			}
+			if err != nil {
+				for i := rng.start + bid; i <= rng.end; i++ {
+					a.blmu.Unlock(a.blockKey(key, i))
+				}
+				if err == io.EOF {
+					break
+				}
+				return nil, err
+			}
+			a.blmu.Unlock(blockID)
+		}
+		return blocks, nil
+	} else if nToFetch == len(blocks) {
 		buf := make([]byte, (rng.end-rng.start+1)*a.blockSize)
 		n, err := a.srcReadAt(key, buf, rng.start*a.blockSize)
```

Comment:
and modify the rest of the calls to only rely on a.srcStreamAt instead of srcReadAt